carma-messenger-core/cpp_message/src/Map_Message.cpp

```diff
@@ -0,0 +1,908 @@
+
+/*
+ * Copyright (C) 2020-2021 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include "Map_Message.h"
+
+namespace cpp_message
+{
+    boost::optional<j2735_msgs::MapData> Map_Message::decode_map_message(std::vector<uint8_t>& binary_array)
+    {
+        j2735_msgs::MapData output;
+
+        // decode results
+        asn_dec_rval_t rval;
+        MessageFrame_t* message = 0;
+        
+         //copy from vector to array         
+        size_t len=binary_array.size();    
+        
+        uint8_t buf[len];             
+        std::copy(binary_array.begin(),binary_array.end(),buf);
+        //use asn1c lib to decode
+        
+        rval=uper_decode(0, &asn_DEF_MessageFrame,(void **) &message, buf, len, 0, 0);
+        if(rval.code == RC_OK)
+        {
+
+
+            auto map_msg = message->value.choice.MapData;
+
+            IntersectionGeometry_t *map_msg_intersections = new IntersectionGeometry_t;
+
+
+            //Intersection Handling
+            if(map_msg.intersections)
```

Comment:
Instead of checking if the new pointer we just created exists, we need to check if the intersections field exists in the incoming message. 
There is a bool "intersections_exists" which we can use here.
The new pointer should be created after that check